### PR TITLE
Improvement: Powder Mining Chat Filter V2

### PR DESCRIPTION
**Pull request by DavidArthurCole**
## Dependencies
Possibly dependent on #2143 due to previous changes being made

## What
Overhaul the Powder Mining filter to allow individual addressing of each type of drop that can come from chests. What was previously a boolean "enabled" now maps to the overall "filter enabled" boolean in the accordion. More complicated options (goblin eggs, gemstones) were given custom enum types for determining functionality.

<details>
<summary>Images</summary>

image

image

</details>

## Changelog Improvements
+ Overhauled the Powder Mining chat filter to be much more customizable. - Daveed

### CODE REVIEW

File: src/main/java/at/hannibal2/skyhanni/features/chat/PowderMiningFilterConfig.java [hunk 1]
```diff
@@ -0,0 +1,114 @@
...
+
+    @Expose
+    @ConfigOption(name = "Wishing Compass", desc = "Hide Wishing Compass rewards.")
+    @ConfigEditorBoolean
+    public boolean wishingCompass = true;
+
+    @Expose
+    @ConfigOption(name = "Oil Barrel", desc = "Hide Oil Barrel rewards.")
+    @ConfigEditorBoolean
+    public boolean oilBarrel = true;
```
Comment (jani270) [inline src/main/java/at/hannibal2/skyhanni/features/chat/PowderMiningFilterConfig.java]: Just a single thought, this here seems a bit ugly and too much, maybe use a list like the Dungeon Filter one?
grafik

Comment (DavidArthurCole) [reply src/main/java/at/hannibal2/skyhanni/features/chat/PowderMiningFilterConfig.java]: Can definitely do that for the boolean ones, gimme 20

Comment (DavidArthurCole) [reply src/main/java/at/hannibal2/skyhanni/features/chat/PowderMiningFilterConfig.java]: image

Fixed in latest commit

**Comment by DavidArthurCole** [comment]
This is completely dead code now that Hypixel changed all their messages.
Guess it's time for me to redo everything.

**Comment by DavidArthurCole** [comment]
image

**Comment by DavidArthurCole** [comment]
This is good to go again. The only slight "nice-to-have" that would be left to do if desired would be re-adding some indication of what the non-filtered results are pertaining to.

Before the formatting change from Hypixel everything had the provided context of "You received..." but now, the unfiltered rewards are just left by themselves in chat.

**Comment by DavidArthurCole** [comment]
Rewards that make it through the filter will now get reformatted to be a lot more readable, and it be much more clear where they came from.
image

**Comment by DavidArthurCole** [comment]
@hannibal002 Just finished my last rounds of verification. It might be worth bumping this up towards 0.26 instead of 0.27, as the current filter is no longer at all functional, and this would be fixing that as well. I can update chglog if necessary.

File: src/main/java/at/hannibal2/skyhanni/features/chat/PowderMiningGemstoneFilterConfig.java [hunk 2]
```diff
@@ -0,0 +1,62 @@
+package at.hannibal2.skyhanni.features.chat;
+
+import com.google.gson.annotations.Expose;
+import io.github.notenoughupdates.moulconfig.annotations.ConfigEditorBoolean;
+import io.github.notenoughupdates.moulconfig.annotations.ConfigEditorDropdown;
+import io.github.notenoughupdates.moulconfig.annotations.ConfigOption;
+
+public class PowderMiningGemstoneFilterConfig {
```
Comment (hannibal002) [inline src/main/java/at/hannibal2/skyhanni/features/chat/PowderMiningGemstoneFilterConfig.java]: wrong package. this belongs into config

File: src/main/java/at/hannibal2/skyhanni/features/chat/PowderMiningChatFilter.kt [hunk 3]
```diff
@@ -0,0 +1,376 @@
...
+                "amethyst" -> gemstoneConfig.amethystGemstones
+                "jade" -> gemstoneConfig.jadeGemstones
+                "topaz" -> gemstoneConfig.topazGemstones
+                //Failure case that should never be reached
+                else -> return "no_filter"
+            }
+
+            if (gemSpecificConfig == GemstoneFilterEntry.HIDE_ALL) return "powder_mining_gemstones"
+
+            when (tierStr) {
```
Comment (hannibal002) [inline src/main/java/at/hannibal2/skyhanni/features/chat/PowderMiningChatFilter.kt]: Move "return" statements from all branches before "when" statement.

File: src/main/java/at/hannibal2/skyhanni/features/chat/PowderMiningChatFilter.kt [hunk 4]
```diff
@@ -0,0 +1,376 @@
...
+     * REGEX-TEST: §r§f❈ Rough Amethyst Gemstone §r§8x24
+     * REGEX-TEST: §r§a❈ Flawed Amethyst Gemstone §r§8x4
+     * REGEX-TEST: §r§9⸕ Fine Amber Gemstone
+     */
+    private val gemstonePattern by patternGroup.pattern(
+        "reward.gemstone",
+        "§r§[fa9][❤❈☘⸕✎✧] (?<tier>Rough|Flawed|Fine) (?<gem>Ruby|Amethyst|Jade|Amber|Sapphire|Topaz) Gemstone( §r§8x(?<amount>[\\d,]+))?",
+    )
+
+    fun block(message: String): String {
```
Comment (hannibal002) [inline src/main/java/at/hannibal2/skyhanni/features/chat/PowderMiningChatFilter.kt:241]: extract this 130 lines function into multiple smaller functions

File: src/main/java/at/hannibal2/skyhanni/features/chat/PowderMiningChatFilter.kt [hunk 5]
```diff
@@ -0,0 +1,376 @@
...
+        )
+        for ((patternToReward, returnReason) in rewardPatterns) {
+            if (patternToReward.first.matches(ssMessage)) {
+                return if (config.simplePowderMiningTypes.contains(patternToReward.second)) returnReason
+                else "no_filter"
+            }
+        }
+
+        //Goblin Eggs
+        goblinEggPattern.matchMatcher(ssMessage) {
```
Comment (hannibal002) [inline src/main/java/at/hannibal2/skyhanni/features/chat/PowderMiningChatFilter.kt:315]: e.g. this matcher's logic can be its own function

File: src/main/java/at/hannibal2/skyhanni/features/chat/PowderMiningChatFilter.kt [hunk 6]
```diff
@@ -0,0 +1,376 @@
...
+            if (config.goblinEggs != PowderMiningFilterConfig.GoblinEggFilterEntry.SHOW_ALL) {
+                if (config.goblinEggs == PowderMiningFilterConfig.GoblinEggFilterEntry.HIDE_ALL) return "powder_mining_goblin_eggs"
+
+                val colorStr = groupOrNull("color")?.lowercase() ?: ""
+                when (colorStr) {
```
Comment (hannibal002) [inline src/main/java/at/hannibal2/skyhanni/features/chat/PowderMiningChatFilter.kt]: Move "return" statements from all branches before "when" statement.

File: src/main/java/at/hannibal2/skyhanni/features/chat/PowderMiningChatFilter.kt [hunk 7]
```diff
@@ -0,0 +1,376 @@
...
+                        "powder_mining_goblin_eggs"
+                    } else "no_filter"
+                    "red" -> return if (config.goblinEggs > PowderMiningFilterConfig.GoblinEggFilterEntry.RED_UP) {
+                        "powder_mining_goblin_eggs"
+                    } else "no_filter"
+                    // BLUE_ONLY enum not explicitly used in comparison, as the only
+                    // case that will block it is HIDE_ALL, which is covered above
+                    "blue" -> return "no_filter"
+                    else -> {
+                        ChatUtils.chat(
```
Comment (hannibal002) [inline src/main/java/at/hannibal2/skyhanni/features/chat/PowderMiningChatFilter.kt]: please use `ErrorManager.logErrorStateWithData()` instead. optionally without the stack trace.

File: src/main/java/at/hannibal2/skyhanni/features/chat/PowderMiningChatFilter.kt [hunk 8]
```diff
@@ -0,0 +1,376 @@
...
+
+        //Gemstones
+        gemstonePattern.matchMatcher(ssMessage) {
+            val gemStr = groupOrNull("gem")?.lowercase() ?: ""
+            val tierStr = groupOrNull("tier")?.lowercase() ?: ""
+
+            //Theoretically impossible but ?
+            if (gemStr.isEmpty() || tierStr.isEmpty()) return ""
+
+            val gemSpecificConfig = when (gemStr) {
```
Comment (hannibal002) [inline src/main/java/at/hannibal2/skyhanni/features/chat/PowderMiningChatFilter.kt]: use a different name. this is not the config, but the setting. the suffix "config" in skyhanni means the actual config element, not the value in it

File: src/main/java/at/hannibal2/skyhanni/features/chat/PowderMiningGemstoneFilterConfig.java [hunk 9]
```diff
@@ -0,0 +1,62 @@
...
+
+    @Expose
+    @ConfigOption(name = "Topaz", desc = "Hide Topaz gemstones under a certain quality.")
+    @ConfigEditorDropdown
+    public GemstoneFilterEntry topazGemstones = GemstoneFilterEntry.FINE_ONLY;
+
+    public enum GemstoneFilterEntry {
+        SHOW_ALL("Show All"),
+        HIDE_ALL("Hide all"),
+        FLAWED_UP("Show §aFlawed§7+"),
```
Comment (hannibal002) [inline src/main/java/at/hannibal2/skyhanni/features/chat/PowderMiningGemstoneFilterConfig.java]: rename the enum name to "flawed or higher" or something similar

**Review by hannibal002 — CHANGES_REQUESTED** [review]
not sure we need options for every different gemstone type. this feels like a waste of config space, and makes setting up stuff complicated for users. Also harder to maintain. (you dont need to remove it, im just curious)

File: src/main/java/at/hannibal2/skyhanni/config/features/chat/FilterTypesConfig.java [hunk 10]
```diff
@@ -51,11 +53,9 @@ public class FilterTypesConfig {
     public boolean winterGift = false;
 
     @Expose
-    @ConfigOption(name = "Powder Mining", desc = "Hide messages while opening chests in the Crystal Hollows. " +
-        "(except powder numbers over 1k, essence numbers over 2, Prehistoric Eggs, and Automaton Parts)")
-    @ConfigEditorBoolean
-    @FeatureToggle
-    public boolean powderMining = false;
+    @ConfigOption(name = "Powder Mining", desc = "")
+    @Accordion
```
Comment (hannibal002) [inline src/main/java/at/hannibal2/skyhanni/config/features/chat/FilterTypesConfig.java]: Please move this accordion either at the beginning or end of the category

**Comment by DavidArthurCole** [comment]
@hannibal002 All requested changes have been addressed in latest commits.

Re: Gemstone config, due to gemstones having different values, seems worth it to have separate configs for each, however if its wanted to flatten that down into one config, I can.

**Review by hannibal002 — APPROVED** [review]

**Pull request merged**